cmd/appliance/upgrade/complete.go

```diff
@@ -455,7 +455,11 @@ func upgradeCompleteRun(cmd *cobra.Command, args []string, opts *upgradeComplete
 						return nil
 					}, backoff.WithContext(backoff.NewExponentialBackOff(), ctx))
 					if err != nil {
-						return fmt.Errorf("upgrade complete failed on %s: %w", i.GetName(), err)
+						if !opts.ciMode {
+							t.Update(appliancepkg.UpgradeStatusFailed)
+							t.Fail(err.Error())
```

Comment:
I have read the code here: https://github.com/appgate/sdpctl/blob/6c4dc6ae53fcd00fd99d4da3f61f534d70a12f06/pkg/tui/tracker.go#L123-L130

I dont fully understand what this does with the select.

We take the error message and place it on the channel continuously until the program ends?

This is the only outcome of calling t.Fail right? Just reading it in this file made me think we were panicking or doing some other exit/fail now behavior but I think it is just a notification message then a sleep.

Reply:
Yep, that's pretty much it. This should never be blocking anything. It's merely for presentation. We've had a lot of issues in the past where these progress bars and trackers were blocking for no apparent reason, so we've ended up with a better-safe-than-sorry approach on these to make sure they never block.